core/store.go

```diff
@@ -8,27 +8,18 @@ import (
 	"github.com/dicedb/dice/server/utils"
 )
 
-type WatchEvent struct {
-	Key       string
-	Operation string
-	Value     *Obj
-}
-
 type Store struct {
 	store        map[string]*Obj
 	expires      map[*Obj]uint64 // Does not need to be thread-safe as it is only accessed by a single thread.
 	keypool      map[string]*string
 	storeMutex   sync.RWMutex
 	keypoolMutex sync.RWMutex
-	WatchList    sync.Map // Maps queries to the file descriptors of clients that are watching them.
-
 }
 
-// WatchChannel Channel to receive updates about keys that are being watched.
-var WatchChannel chan WatchEvent
-
 func NewStore() *Store {
-	WatchChannel = make(chan WatchEvent, config.KeysLimit)
+	WatchChan = make(chan WatchEvent, config.KeysLimit)
+	WatchSubscriptionChan = make(chan WatchSubscription)
```

Comment:
This logic should likely be decoupled from the Store initialisation. But it's probably a better idea to do it once we have a more concrete structure for the shared nothing Architecture in place.

Reply:
We will likely only have a single query watcher for the entire system (instead of it being shard-specific). Each shard publishes a message on the same WatchChan which the watcher listens on and runs queries on (the queries run across shards instead of on a single shard)

Not sure if we gain much from having multiple of these.

Reply:
> We will likely only have a single query watcher for the entire system (instead of it being shard-specific
Wouldn't this need to broken down into shard specific watchers as the watch event will be tightly coupled to a store?

May be i am missing a point here. We can discuss this over a call...

Reply:
A watch event originates from a single shard (and is coupled with it). But this event affects a query and that query needs to be completely re-evaluated.

So when shardA sends a watch event, it reaches the central QueryWatcher. This watcher holds all the information about which clients are watching which queries. If this watch eveny affects any of these queries it launches a query execution (which happens across all shards)

Therefore, even if a watch event is coupled with a single shard, the query must be evaluated across all shards.

Reply:
Thanks for sharing your thoughts. It is brings in new perspectives for me and forcing me to think further. I am loving this discussion.

> So when shardA sends a watch event, it reaches the central QueryWatcher.

I am trying to understand the role of CentralQueryWatcher a little better.
So when a IOThread(WorkerThread) executes a QWATCH query, would it send the request to the CentralQueryWatcher which co-ordinates the whole thing and sends back the updates to the IOThread(WorkerThread)?
Wouldn't this require us to have extra co-ordination?

> This watcher holds all the information about which clients are watching which queries. If this watch eveny affects any of these queries it launches a query execution (which happens across all shards)

What do you think about moving the CentralQueryWatcher inside making it scoped to individual shards.
This way we can avoid a CentralQueryWatcher to deal with individual store events from each of the shards and becoming a bottleneck?

Reply:
Also this can align with general scatter-gather approach typical to each eval behaviour implementation we were discussing the other day.

Reply:
1. When a client sends a QWATCH request, it registers itself with the query watcher (the watcher adds it to the WatchList). It then proceeds to run a query immediately to return the initial result.
2. Later, when updates are delivered to the QueryWatcher it can automatically re-trigger queries and deliver results to the client.

Keeping the query watcher inside each shard feels restrictive. Whenever the watcher would want to execute a query it would need to lock all the other shards (for now). This behaviour is much simpler to reason about if this watcher exists as a separate entity from that shards (instead of being coupled with one shard and trying to acquire a lock on all others).

Further, we had plans that each query could maintain query specific shards of keys which the query may be interested in, to improve performance. This behaviour becomes a little harder to co-ordinate if we have multiple QueryWatchers (where each watcher maintains sub shards for each query).
A central watcher can allow us to maintain such query specific shards in one location.

Finally, I agree it may be better if we can model this in a scatter-gather approach, but I don't think implementing such an approach is easy to do with queries (the gathering logic becomes highly specific for each query, and therefore becomes error prone). A central query Executor allows us to keep the query execution logic a lot simpler.

Reply:
Thanks for the time and patience for explaining in detail. I now understand. Given the generic nature of the query, a central query engine(enabling watch/range based queries) makes a lot of sense.

May be to handle such cases we can have a centralized B+Tree or LSM based ReadOnly DB(that contains all the data across all shards) like DragonFly/Redis does?

Reply:
I agree, a specialised structure for serving queries is a good idea. We can start working on these optimizations once we have the e2e flow working.